masp_primitives/src/transaction.rs

```diff
@@ -0,0 +1,681 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use core::fmt::Debug;
+use ff::PrimeField;
+use group::GroupEncoding;
+use std::collections::BTreeMap;
+use std::convert::TryInto;
+use std::io::{self, Read, Write};
+
+use masp_note_encryption::{
+    EphemeralKeyBytes, ShieldedOutput, COMPACT_NOTE_SIZE, ENC_CIPHERTEXT_SIZE,
+};
+
+use crate::{
+    asset_type::AssetType,
+    consensus::{self, BranchId},
+    note_encryption::SaplingDomain,
+    primitives::Nullifier,
+    prover::GROTH_PROOF_SIZE,
+    redjubjub::{self, PublicKey, Signature},
+    transaction::amount::Amount,
+};
+
+pub mod amount;
+pub mod builder;
+pub mod memo;
+pub mod serialize;
+pub mod util;
+
+pub type GrothProofBytes = [u8; GROTH_PROOF_SIZE];
+const SAPLING_VERSION_GROUP_ID: u32 = 0x892F2085;
+const SAPLING_TX_VERSION: u32 = 4;
+
+/// The set of defined transaction format versions.
+///
+/// This is serialized in the first four or eight bytes of the transaction format, and
+/// represents valid combinations of the `(overwintered, version, version_group_id)`
+/// transaction fields. Note that this is not dependent on epoch, only on transaction encoding.
+/// For example, if a particular epoch defines a new transaction version but also allows the
+/// previous version, then only the new version would be added to this enum.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum TxVersion {
```

Comment:
It seems that we don't need `TxVersion` here. It's the ledger's responsibility to define tx version?

Reply:
Good point, let's get rid of it